cloud/aws/providerconfig/types.go

```diff
@@ -14,22 +14,156 @@
 package providerconfig
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
+// for an AWS instance.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderConfig struct {
 	metav1.TypeMeta `json:",inline"`
+
+	// AMI is the reference to the AMI from which to create the machine instance.
+	AMI AWSResourceReference
+
+	// InstanceType is the type of instance to create. Example: m4.xlarge
+	InstanceType string
+
+	// Tags is the set of tags to add to apply to an instance, in addition to the ones
+	// added by default by the actuator. These tags are additive. The actuator will ensure
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
+
+	// IAMInstanceProfile is a reference to an IAM role to assign to the instance
+	IAMInstanceProfile *AWSResourceReference
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
+
+	// CredentialsSecret is a reference to the secret with AWS credentials. Otherwise, defaults to permissions
+	// provided by attached IAM role where the actuator is running.
+	CredentialsSecret *corev1.LocalObjectReference
+
+	// KeyName is the name of the KeyPair to use for SSH
+	KeyName *string
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
+
+	// PublicIP specifies whether the instance should get a public IP. If not present,
+	// it should use the default of its subnet.
+	PublicIP *bool
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+
+	// Placement specifies where to create the instance in AWS
+	Placement Placement
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it is created.
+	LoadBalancerIDs []string
+}
+
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters.
+// Only one of ID, ARN or Filters may be specified. Specifying more than one will result in
+// a validation error.
+type AWSResourceReference struct {
```

Comment:
This seems like it could get messy? How does this fit into the MVP?

Reply:
It allows you to specify references to things that might not exist yet when you create your machine/machineset config. For example, you may know what your security groups will be named, but you don't have id's for them yet. Your config can use a Filter for the name tag for each of your security groups.

Reply:
So far all of the resources that we reference from the machine config have a corresponding Describe call that allows them to be looked up by the fields specified in the struct. I would say we could introduce separate types of references in the future if/when we need them. Is that ok?

Reply:
Sounds good, thanks for the clarification.